### PR TITLE
Subject Assistant Exports - Prototype

### PR DESCRIPTION
**Pull request by shaunanoordin**
## PR Overview

Connected Project: [Zoo ML Subject Assistant](https://github.com/zooniverse/zoo-ml-subject-assistant)

This PR adds functionality required by the Subject Assistant web app. The features added will allow Hamlet to create a  "Subject Manifest" that will be uploaded to an external Machine Learning service, and that ML service will then be used by the web app for further processing.

PR Goals:
- [x] a webpage should display a list of all Subject Sets for a given project, e.g. `http://localhost:8080/subject-assistant/1292/`
  - [x] For each Subject Set, there should be a "Generate Subject Manifest" button
  - [x] The button should be linked to an API that generates a Subject Manifest file, e.g. `http://localhost:8080/subject-assistant/1292/subject-sets/78138/`
    - (File should be stored... somewhere... for future uploading to an external service.)
  - [x] there should be a status indicating whether the Subject Manifest has been created.

NOT PR Goals:
- Uploading the Subject Manifest to an external ML service to perform an Machine Learning Task.
- Retrieving the "Task ID" returned by the ML service

Notes:
- The Subject Assistant Exports could benefit from specifying WHICH ML service to export to, in case multiple crop up in the future.
- The Subject Manifest file should look something like this:
```
[
  [
    "https://panoptes-uploads.zooniverse.org/staging/subject_location/3eb45309-ad43-4819-abf1-59bf083e8154.jpeg",
    "{\"project_id\":\"1891\",\"workflow_id\":\"3352\",\"subject_set_id\":\"78138\",\"subject_id\":\"83481\",\"frame_id\":\"0\"}" ],
  [...],
  ...
]
```
- Basically: `Array(Array(String:image_url, String:stringified_json_metadata ))` where `stringified_json_metadata` should contain the Project ID, SS ID, Workflow ID, Subject ID, and etc that will be important for the Subject Assistant web app to reconstitute a Subject.

### Status

WIP

### CODE REVIEW

**Comment by shaunanoordin** [comment]
Current challenges:
- POSTing to `http://localhost:8080/subject-assistant/1292/subject-sets/78138/` current results in:
  ```
  relation "exports_mlsubjectassistantexport" does not exist
  LINE 1: INSERT INTO "exports_mlsubjectassistantexport" ("status", "c...
  ```
  - So this means I haven't fully grasped how to set up the database table in Django. 
  - (`docker-compose up` again maybe?)
- I'm not sure what's performing the "convert Subject Set resource into a CSV" in the standard Subject Sets Export example.
  - It's... a... Celery task? Because... Celery's job is to run async tasks like Sidekiq or web workers... and... therefore... those tasks need to be defined somewhere?
  - Maybe `/hamlet/celery.py`?
 
Optimisation:
- I'm thinking I should separate the Subject Assistant code into separate files, e.g. `(SA code inside) views.py -> subject_assistant/views.py` but I should check whether Django/this repo's code should be organised by function/actions or by modules.

**Comment by adammcmaster** [comment]
> Current challenges:
> 
> * POSTing to `http://localhost:8080/subject-assistant/1292/subject-sets/78138/` current results in:
>   ```
>   relation "exports_mlsubjectassistantexport" does not exist
>   LINE 1: INSERT INTO "exports_mlsubjectassistantexport" ("status", "c...
>   ```
>   
>   
>   
>   * So this means I haven't fully grasped how to set up the database table in Django.
>   * (`docker-compose up` again maybe?)

You'll need to generate a new database migration and then run it (you'll generally need to do this whenever you make changes to models):

```
docker-compose run app python manage.py makemigrations
docker-compose run app python manage.py migrate
```

> * I'm not sure what's performing the "convert Subject Set resource into a CSV" in the standard Subject Sets Export example.
>   
>   * It's... a... Celery task? Because... Celery's job is to run async tasks like Sidekiq or web workers... and... therefore... those tasks need to be defined somewhere?
>   * Maybe `/hamlet/celery.py`?

Yup, at the moment all the background tasks are defined in that one file. That might get confusing as we add more, so it might be worth thinking about refactoring at some point.

> 
> Optimisation:
> 
> * I'm thinking I should separate the Subject Assistant code into separate files, e.g. `(SA code inside) views.py -> subject_assistant/views.py` but I should check whether Django/this repo's code should be organised by function/actions or by modules.

That makes sense to me. The normal way to organise things is by module/app.

**Comment by shaunanoordin** [comment]
## PR Overview

This branch is now feature-complete and ready for review.

Features:
- The Subject Assistant export page is now accessible by going to `/subject-assistant/${project_id}`, e.g. `http://localhost:8080/subject-assistant/1292/`
  - This page will list all the Subject Sets for a given project.
  - Each Subject Set has a 'Generate' button that creates the "Subject Manifest" JSON to be exported
  - Each Subject Set has a status indicator and, when successful, a link to the generated JSON file.

### Full Dev Notes

Code changes, database:
- A new model, `MLSubjectAssistantExport`, has been created with two points of data: the subject set's ID, and the generated JSON file (along with its processing status).

Code changes, Django:
- New path: `GET /subject-assistant/${project_id}/` will list all the Subject Sets for a Project. (HTML page)
- New path: `POST /subject-assistant/${project_id}/subject-sets/${subject_set_id}/` will trigger the 'generate JSON export' task. (then redirects to 'list Subject Sets' page)

Code changes, Celery:
- new 'generate JSON export' task created: `ml_subject_assistant_export_to_microsoft()`
  - Asynchronous. Requires connection to Zooniverse API and Zoo authentication. ⚠️ Requires file to be written to a temporary directory (see above: '/tmp' needs to be created at the root folder of the project for local testing)
  - This task accepts a Subject Set ID as an argument.
  - This task outputs a JSON file, saved to the database.

Room for improvement: (next PR?)
- documentation for the Subject Assistants feature
- slicing the code and placing them into a Subject Assistant module folder (once I figure Python imports without crashing stuff again)
- inline comments for Subject Assistants `views`.
- ⚠️  `ml_subject_assistant_export()` action SHOULD HAVE A SAFETY to prevent it from being run if it's already running. We don't want to accidentally DDOS ourselves or Microsoft by letting people click that button multiple times.

### Status

Ready for review 👌

File: hamlet/celery.py [hunk 1]
```diff
@@ -277,3 +278,97 @@ def workflow_export(
...
+# Creates a JSON file which details all the image URLs in a Subject Set, along
+# with that image's associated Subject information. JSON will be of the format:
+#
+# [ [ "http://image.url/example.png",
+#     "{ \"subject_info\": \"is a stringified JSON object\" }" ],
+#   ... ]
+#
+# To be used with the Zooniverse ML Subject Assistant web app (https://github.com/zooniverse/zoo-ml-subject-assistant),
+# and the associated (and currently unnamed) Microsoft machine learning service.
+# ------------------------------------------------------------------------------
```
Comment (adammcmaster) [inline hamlet/celery.py]: Could you move this comment into a docstring? (See here for an example https://www.python.org/dev/peps/pep-0257/#multi-line-docstrings)

File: hamlet/celery.py [hunk 2]
```diff
@@ -277,3 +278,97 @@ def workflow_export(
...
+        export.status = MLSubjectAssistantExport.RUNNING
+        export.save()
+        with SocialPanoptes(bearer_token=access_token) as p:
+            subject_set = SubjectSet.find(export.subject_set_id)
+            
+            # Process each Subject
+            for subject in subject_set.subjects:
+                
+                # Create a data item for each image URL in the Subject
+                frame_id = 0
```
Comment (adammcmaster) [inline hamlet/celery.py]: I'd consider using the `enumerate` function here:

```python
for frame_id, location in enumerate(subject.locations):
```

File: hamlet/celery.py [hunk 3]
```diff
@@ -277,3 +278,97 @@ def workflow_export(
...
+                        'frame_id': str(frame_id)
+                    }
+                    
+                    item = []
+                    item.append(image_url)
+                    item.append(json.dumps(subject_information))  # The subject's JSON information is stored as a string. Yes, really.
+                    
+                    data.append(item)
+                    
+                    frame_id += 1
```
Comment (adammcmaster) [inline hamlet/celery.py]: You can remove this line if you use `enumerate` above.

File: hamlet/celery.py (hunk 1 above)
Comment (shaunanoordin) [reply hamlet/celery.py]: 👌

File: hamlet/celery.py (hunk 2 above)
Comment (shaunanoordin) [reply hamlet/celery.py]: 👌

**Comment by shaunanoordin** [comment]
## PR Update

Thanks for the review @adammcmaster ! I've made updates:

- Comments for functions now use the Python docstring method, and
- The 'for' loop now uses Pythong's `enumerate()`

👍 I'm adding those two items to my Python toolbox.
